aten/src/ATen/DeviceGuard.h

```diff
@@ -6,131 +6,180 @@
+  /// Set the current device to the passed Device, if not nullopt;
+  /// otherwise do nothing.
+  explicit DeviceGuard(optional<Device> device_opt) {
     if (device_opt.has_value()) {
-      set_device(device_opt.value());
+      init_device(device_opt.value());
     }
   }
 
   /// Sets the current device to the device on which the given tensor is located.
   explicit DeviceGuard(const Tensor& tensor) {
-    set_device_from(tensor);
+    init_device_from(tensor);
   }
 
   /// Sets the current device to the device on which the first tensor in the list is
   /// located. If the list is empty, does nothing.
   explicit DeviceGuard(const TensorList& tensors) {
     if (!tensors.empty()) {
-      set_device_from(tensors.front());
+      init_device_from(tensors.front());
     }
   }
 
+  /// A constructor for testing; permits explicitly passing in the
+  /// DeviceGuardImpl.
+  explicit DeviceGuard(Device device, const detail::DeviceGuardImplInterface* impl) {
+    init_device(device, impl);
+  }
+
   /// Copy is disallowed.
   DeviceGuard(const DeviceGuard&) = delete;
   DeviceGuard& operator=(const DeviceGuard&) = delete;
 
-  /// Move-constructs this `DeviceGuard` from another `DeviceGuard`. The
-  /// moved-from `DeviceGuard` is modified such that its destruction has no
-  /// effect (does not reset the device).
+  /// Move-constructs this `DeviceGuard` from another `DeviceGuard`.
+  /// This can be used to terminate the lifetime of a `DeviceGuard`
+  /// early; for example, in:
+  ///
+  ///     // current device is d0
+  ///     DeviceGuard g1(d1);
+  ///     // current device is d1
+  ///     {
+  ///       DeviceGuard g2(std::move(g1));
+  ///     }
+  ///     // current device is d0!!
+  ///
+  /// It is undefined behavior if you move a device guard across
+  /// an intervening DeviceGuard.
+  ///
+  ///     DeviceGuard g1(d1);
+  ///     DeviceGuard g2(d2);
+  ///     DeviceGuard g3(std::move(g1)); // UB!
+  ///
   DeviceGuard(DeviceGuard&& other) noexcept {
-    *this = std::move(other);
+    // Default construction leaves this uninitialized
+    std::swap(impl_, other.impl_);
+    std::swap(original_device_, other.original_device_);
+    std::swap(current_device_, other.current_device_);
   }
 
-  /// Move-assigns this `DeviceGuard` from another `DeviceGuard`. The
-  /// moved-from `DeviceGuard` is modified such that its destruction has no
-  /// effect (does not reset the device).
+  /// Move-assigns this `DeviceGuard` from another `DeviceGuard`. This
+  /// `DeviceGuard` is immediately terminated.  This allows
+  /// you to implement a modest performance optimization: if the device
+  /// type of the previous DeviceGuard and the new DeviceGuard match,
+  /// then skips the otherwise unnecessary setDevice from the previous
+  /// device guard.
+  ///
+  /// As with the move constructor, it is undefined behavior if you
+  /// move a device guard across an intervening DeviceGuard.
+  ///
```

Comment:
I thought you decided to get rid of move assignment

Reply:
Yeah that's weird, I thought I did remove it. Checking my diffs...

Reply:
Oops, I lost this hunk somehow. Updating.